js/plugin-loader.js

```diff
@@ -43,113 +44,75 @@ const defaultOptions = {
   serverConfig: null
 }
 
-function ExternalImports() {
-  this.map = {};
+function Service(def, configuration, plugin) {
+  this.configuration = configuration;
+  //don't do this here: avoid circular structures:
+  //this.plugin = plugin; 
+  Object.assign(this, def);
+  if (this.version === undefined) {
+    //FIXME temporary hack until everyone has fixed their plugin defs.
+    bootstrapLogger.warn(`Service ${plugin.identifier}::${this.name} doesn't` 
+      + ` specify a version. Resorting to the plugin's API version ${plugin.apiVersion}`);
+    this.version = plugin.apiVersion;
+  }
 }
-
-ExternalImports.prototype = {
-  constructor: ExternalImports,
-  /**
-   * A plugin import table in the opposite direction. 
-   * Shows who imports us, rather than whom we import.
-   *   
-   *   <plugin id>: {
-   *     <service name>: [
-   *       ...,
-   *       {
-   *         "toPlugin": the id of a plugin that imports the service from the plugin
-   *         "alias": the target alias of the service
-   *       },
-   *       ...
-   *     ]
-   *   }
-   */
-  map: null, 
+Service.prototype = {
+  constructor: Service,
   
-  /**
-   * Register a newly obtained plugin definition's dependencies to the table
-   */
-  registerAsImporterOfAs(destPlugin, sourcePlugin, sourceName, localName) {
-    let sourcePluginEntry = this.map[sourcePlugin];
-    if (!sourcePluginEntry) {
-      sourcePluginEntry = this.map[sourcePlugin] = {};
-    }
-    let sourceServiceRefs = sourcePluginEntry[sourceName];
-    if (!sourceServiceRefs) {
-      sourceServiceRefs = sourcePluginEntry[sourceName] = [];
+  validate() {
+    if (!semver.valid(this.version)) {
+      throw new Error(`${this.name}: invalid version "${this.version}"`)
     }
-    sourceServiceRefs.push({
-      toPlugin: destPlugin.identifier,
-      alias: localName
-    });
-  },
-  
-  /**
-   * Ensure that none of the `plugins` contains unsatisfied imports.
-   * 
-   * Every service of every plugin from `plugins` cancels out the corresponding
-   * records from the table.
-   * 
-   * Updates every imported data service's externalRefs
-   * 
-   * Returns true if the table ends up being empty, false otherwise
-   */
-  allImportsResolved(plugins) {
-    let success = true;
-    for (const plugin of plugins) {
-      const externalImports = this.map[plugin.identifier];
-      if (externalImports) {
-        if (plugin.dataServices) {
-          for (const dataservice of plugin.dataServices) {
-            dataservice.externalRefs = externalImports[dataservice.name];
-            delete externalImports[dataservice.name];
-          }
+    if (this.versionRequirements) {
```

Comment:
See your own comment here, notice the "requiredServices" property (I slightly reorganized/rephrased it but the idea is the same)
https://jira.rocketsoftware.com/browse/MVD-206?focusedCommentId=7152355&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-7152355

Reply:
You must have confused me back then because I'm confused now. Version requirement marking is important for imports, but for other dataservice types it seems that while you can make this relationship statement it wouldn't add value. A plugin is developed to work as a coherent unit, so doing version checks from one service to another in the same plugin seems pointless in terms of the compatibility already being guaranteed by the developer as a result of publishing the plugin.

Let's say, for example, that service foo had its API updated, such that the new service was able to serve APIs v1 and v2. And let's say that service bar in the same plugin relies on calling foo @ v1. Does this relationship really need to be stated? Isn't it implied that a given release of a plugin is valid by the developer releasing it as a whole, rather than in parts? I mean, if bar needed v1 but foo only provided v2, it would not work but specifying that additional metadata wouldn't solve the problem. People would just issue a bug report which would lead to the fix instead.

However - you do of course need to do version requirement checks for IMPORT dataservices, since you should only be able to import when the version pattern matches, since a given plugin expects a service of a certain version, which is obtained externally and therefore has a higher chance of failure due to out of date plugins.

So, for an import, imagine:
    {
      "version": "//makes sense to be inherited by the import resolution",
      "type": "import",
      "sourceName": "hello",
      "localName": "hello",
      "sourcePlugin": "org.zowe.zlux.sample.iframe",
      "versionRequirements": "~3.0.0"
    },

Reply:
Considering this scenario: there's service A in two versions: v1 and v2, and there's service B who wants to call A, not the most recent version, but v1 instead.  I don't find this scenario impossible. I don't see why this is necessarily a bug: what if they deliberately don't want to update B to the newer version of A's API for some reason? Furthermore, if in the situation of B calling A we just choose the most recent  version of A, that would simply be an arbitrary choice on our part: why do we think that newest is the best? 

That was exactly the question back then: should we actually bother to implement those "local version requirements" or not? I thought it's worth it, but if you think that's useless, probably maintaining that code would be a burden and I'd need to remove it.

Reply:
OK, you raise a good point about versioning for non-import services but I'm having trouble finding the code that could do this: if B depends on a version of A, callService should either have a version parameter or know that a named service A requested by B corresponds to a particular version of A due to the stated versionRequirements. There's an injectible middleware piece for associating proxies with callService but it doesn't look like it has changed to accommodate versioning.

Reply:
I don't get your question: one part is done in the code we're discussing right here, the other part is indeed in callService(): https://github.com/zowe/zlux-proxy-server/pull/14/files#diff-55d2935735e0c392902cca19a95e92c0R299